File: tests/oracle/varmap_gather_offset_1204.cc
#include "op.hh"
#include "typeop.hh"
#include "varmap.hh"
#include "varnode.hh"

#include <iomanip>
#include <iostream>
#include <iterator>
#include <string>
#include <vector>

namespace {

using namespace ghidra;

struct OpSetInputTag {
  using type = void (PcodeOp::*)(Varnode *, int4);
  friend type access(OpSetInputTag);
};

struct OpSetOutputTag {
  using type = void (PcodeOp::*)(Varnode *);
  friend type access(OpSetOutputTag);
};

struct VarnodeSetDefTag {
  using type = void (Varnode::*)(PcodeOp *);
  friend type access(VarnodeSetDefTag);
};

struct VarnodeAddDescendTag {
  using type = void (Varnode::*)(PcodeOp *);
  friend type access(VarnodeAddDescendTag);
};

template <typename Tag, typename Tag::type Member>
struct PrivateAccess {
  friend typename Tag::type access(Tag) { return Member; }
};

template struct PrivateAccess<OpSetInputTag, &PcodeOp::setInput>;
template struct PrivateAccess<OpSetOutputTag, &PcodeOp::setOutput>;
template struct PrivateAccess<VarnodeSetDefTag, &Varnode::setDef>;
template struct PrivateAccess<VarnodeAddDescendTag, &Varnode::addDescend>;

size_t countDescendants(const Varnode &vn) {
  return static_cast<size_t>(std::distance(vn.beginDescend(), vn.endDescend()));
}

void dumpState(const std::string &label, const char *phase, const Varnode &output,
               const PcodeOp *op, const std::vector<Varnode *> &inputs) {
  const Datatype *outputType = output.getType();
  std::string outputMetatype;
  metatype2string(outputType->getMetatype(), outputMetatype);
  std::cout << label << '|' << phase
            << "|out_unique=" << (output.getSpace()->getType() == IPTR_INTERNAL)
            << "|out_offset=" << output.getOffset()
            << "|out_size=" << output.getSize()
            << "|out_flags=" << output.getFlags()
            << "|out_consumed=" << output.getConsume()
            << "|out_nzm=" << output.getNZMask()
            << "|def_alias=" << (output.getDef() == op)
            << "|output_alias=" << (op->getOut() == &output)
            << "|out_type_present=" << (outputType != nullptr)
            << "|out_type_name=" << outputType->getName()
            << "|out_type_size=" << outputType->getSize()
            << "|out_type_metatype=" << outputMetatype
            << "|opcode=" << static_cast<int4>(op->code())
            << "|inputs=" << op->numInput()
            << "|dead=" << op->isDead()
            << "|eval=" << op->getEvalType()
            << "|commutative=" << op->isCommutative()
            << "|seq_offset=" << op->getAddr().getOffset()
            << "|seq_order=" << op->getTime();
  for (size_t i = 0; i < inputs.size(); ++i) {
    const Varnode *input = inputs[i];
    const Datatype *inputType = input->getType();
    std::string inputMetatype;
    metatype2string(inputType->getMetatype(), inputMetatype);
    std::cout << "|in" << i << "_constant=" << input->isConstant()
              << "|in" << i << "_offset=" << input->getOffset()
              << "|in" << i << "_size=" << input->getSize()
              << "|in" << i << "_flags=" << input->getFlags()
              << "|in" << i << "_consumed=" << input->getConsume()
              << "|in" << i << "_nzm=" << input->getNZMask()
              << "|in" << i << "_slot_alias=" << (op->getIn(i) == input)
              << "|in" << i << "_descendants=" << countDescendants(*input)
              << "|in" << i << "_type_present=" << (inputType != nullptr)
              << "|in" << i << "_type_name=" << inputType->getName()
              << "|in" << i << "_type_size=" << inputType->getSize()
              << "|in" << i << "_type_metatype=" << inputMetatype
              << "|in" << i << "_type_alias_out=" << (inputType == outputType);
  }
  if (inputs.size() == 2) {
    std::cout << "|input_alias=" << (inputs[0] == inputs[1])
              << "|input_type_alias=" << (inputs[0]->getType() == inputs[1]->getType());
  }
  std::cout << '\n';
}

void runCopy(const std::string &label, int4 size, uintb value) {
  ConstantSpace constantSpace(nullptr, nullptr);
  UniqueSpace uniqueSpace(nullptr, nullptr);
  TypeBase valueType(size, TYPE_UNKNOWN, "fixture_unknown");
  TypeOpCopy copyType(nullptr);
  PcodeOpBank bank;
  PcodeOp *op = bank.create(1, Address(&uniqueSpace, 0x1000));
  bank.changeOpcode(op, &copyType);
  Varnode input(size, Address(&constantSpace, value), &valueType);
  Varnode output(size, Address(&uniqueSpace, 0x2000), &valueType);
  const auto setInput = access(OpSetInputTag{});
  const auto setOutput = access(OpSetOutputTag{});
  const auto setDef = access(VarnodeSetDefTag{});
  const auto addDescend = access(VarnodeAddDescendTag{});
  (op->*setInput)(&input, 0);
  (op->*setOutput)(&output);
  (input.*addDescend)(op);
  (output.*setDef)(op);
  const std::vector<Varnode *> inputs{&input};

  dumpState(label, "before", output, op, inputs);
  const uintb result = AliasChecker::gatherOffset(&output);
  std::cout << label << "|result=0x" << std::hex << result << std::dec << '\n';
  dumpState(label, "after", output, op, inputs);
}

void runAdd(const std::string &label, int4 size, uintb left, uintb right) {
  ConstantSpace constantSpace(nullptr, nullptr);
  UniqueSpace uniqueSpace(nullptr, nullptr);
  TypeBase valueType(size, TYPE_UNKNOWN, "fixture_unknown");
  TypeOpIntAdd addType(nullptr);
  PcodeOpBank bank;
  PcodeOp *op = bank.create(2, Address(&uniqueSpace, 0x1100));
  bank.changeOpcode(op, &addType);
  Varnode input0(size, Address(&constantSpace, left), &valueType);
  Varnode input1(size, Address(&constantSpace, right), &valueType);
  Varnode output(size, Address(&uniqueSpace, 0x2100), &valueType);
  const auto setInput = access(OpSetInputTag{});
  const auto setOutput = access(OpSetOutputTag{});
  const auto setDef = access(VarnodeSetDefTag{});
  const auto addDescend = access(VarnodeAddDescendTag{});
  (op->*setInput)(&input0, 0);
  (op->*setInput)(&input1, 1);
  (op->*setOutput)(&output);
  (input0.*addDescend)(op);
  (input1.*addDescend)(op);
  (output.*setDef)(op);
  const std::vector<Varnode *> inputs{&input0, &input1};

  dumpState(label, "before", output, op, inputs);
  const uintb result = AliasChecker::gatherOffset(&output);
  std::cout << label << "|result=0x" << std::hex << result << std::dec << '\n';
  dumpState(label, "after", output, op, inputs);
}

} // namespace

int main() {
  runCopy("copy8_all_bits", 8, 0xfedcba9876543210ULL);
  runAdd("add8_wrap", 8, 0xfffffffffffffff0ULL, 0x35);
  runAdd("add7_mask", 7, 0x00fffffffffffff0ULL, 0x35);
  return 0;
}

File: tests/oracle/varmap_gather_offset_1204.rs
use std::sync::{Arc, RwLock};

use rugra::address::{Address, SeqNum};
use rugra::op::{pcodeop_flags, PcodeOp};
use rugra::opcodes::OpCode;
use rugra::space::AddressSpace;
use rugra::type_system::datatype::{Datatype, TypeBase, TypeMetatype};
use rugra::varmap::gather_offset;
use rugra::varnode::{varnode_flags, Varnode};

fn count_descendants(vn: &Arc<RwLock<Varnode>>) -> usize {
    vn.read()
        .unwrap()
        .descend
        .iter()
        .filter(|weak| weak.upgrade().is_some())
        .count()
}

fn dump_state(
    label: &str,
    phase: &str,
    output: &Arc<RwLock<Varnode>>,
    op: &Arc<RwLock<PcodeOp>>,
    inputs: &[Arc<RwLock<Varnode>>],
) {
    let output_guard = output.read().unwrap();
    let op_guard = op.read().unwrap();
    let output_type = output_guard.get_type().expect("fixture output type");
    assert_eq!(output_type.get_metatype(), TypeMetatype::Unknown);
    print!(
        "{label}|{phase}|out_unique={}|out_offset={}|out_size={}|out_flags={}|out_consumed={}|out_nzm={}|def_alias={}|output_alias={}|out_type_present=1|out_type_name={}|out_type_size={}|out_type_metatype=unknown|opcode={}|inputs={}|dead={}|eval={}|commutative={}|seq_offset={}|seq_order={}",
        u8::from(output_guard.is_unique()),
        output_guard.get_offset(),
        output_guard.get_size(),
        output_guard.flags,
        output_guard.get_consume(),
        output_guard.get_nzm(),
        u8::from(
            output_guard
                .get_def()
                .map(|definition| Arc::ptr_eq(&definition, op))
                .unwrap_or(false)
        ),
        u8::from(
            op_guard
                .output
                .as_ref()
                .map(|candidate| Arc::ptr_eq(candidate, output))
                .unwrap_or(false)
        ),
        output_type.get_name(),
        output_type.get_size(),
        op_guard.opcode as u32,
        op_guard.inrefs.len(),
        u8::from(op_guard.is_dead()),
        op_guard.get_eval_type(),
        u8::from((op_guard.flags & pcodeop_flags::COMMUTATIVE) != 0),
        op_guard.get_addr().as_u64(),
        op_guard.get_seq_num().get_order(),
    );
    for (index, input) in inputs.iter().enumerate() {
        let input_guard = input.read().unwrap();
        let input_type = input_guard.get_type().expect("fixture input type");
        assert_eq!(input_type.get_metatype(), TypeMetatype::Unknown);
        print!(
            "|in{index}_constant={}|in{index}_offset={}|in{index}_size={}|in{index}_flags={}|in{index}_consumed={}|in{index}_nzm={}|in{index}_slot_alias={}|in{index}_descendants={}|in{index}_type_present=1|in{index}_type_name={}|in{index}_type_size={}|in{index}_type_metatype=unknown|in{index}_type_alias_out={}",
            u8::from(input_guard.is_constant()),
            input_guard.get_offset(),
            input_guard.get_size(),
            input_guard.flags,
            input_guard.get_consume(),
            input_guard.get_nzm(),
            u8::from(Arc::ptr_eq(&op_guard.inrefs[index], input)),
            count_descendants(input),
            input_type.get_name(),
            input_type.get_size(),
            u8::from(Arc::ptr_eq(&input_type, &output_type)),
        );
    }
    if inputs.len() == 2 {
        let input0_type = inputs[0].read().unwrap().get_type().unwrap();
        let input1_type = inputs[1].read().unwrap().get_type().unwrap();
        print!(
            "|input_alias={}|input_type_alias={}",
            u8::from(Arc::ptr_eq(&inputs[0], &inputs[1])),
            u8::from(Arc::ptr_eq(&input0_type, &input1_type)),
        );
    }
    println!();
}

fn make_op(opcode: OpCode, pc: u64, inputs: &[Arc<RwLock<Varnode>>]) -> Arc<RwLock<PcodeOp>> {
    let mut op = PcodeOp::new(SeqNum::new(Address::new(pc), 0), opcode);
    op.flags = pcodeop_flags::DEAD
        | match opcode {
            OpCode::CPUI_COPY => pcodeop_flags::UNARY | pcodeop_flags::NOCOLLAPSE,
            OpCode::CPUI_INT_ADD => pcodeop_flags::BINARY | pcodeop_flags::COMMUTATIVE,
            _ => unreachable!("fixture only constructs COPY and INT_ADD"),
        };
    op.inrefs = inputs.to_vec();
    Arc::new(RwLock::new(op))
}

fn attach(
    size: usize,
    offset: u64,
    opcode: OpCode,
    pc: u64,
    datatype: &Arc<Datatype>,
    inputs: &[Arc<RwLock<Varnode>>],
) -> (Arc<RwLock<Varnode>>, Arc<RwLock<PcodeOp>>) {
    let output = Arc::new(RwLock::new(Varnode::new_with_space(
        size,
        AddressSpace::Unique,
        offset,
    )));
    let op = make_op(opcode, pc, inputs);
    {
        let mut output_guard = output.write().unwrap();
        output_guard.set_flags(varnode_flags::WRITTEN | varnode_flags::COVERDIRTY);
        output_guard.def = Some(Arc::downgrade(&op));
        output_guard.v_type = Some(datatype.clone());
        output_guard.consumed = u64::MAX;
        output_guard.nzm = u64::MAX;
    }
    op.write().unwrap().output = Some(output.clone());
    for input in inputs {
        input.write().unwrap().add_descend(&op);
    }
    (output, op)
}

fn constant(value: u64, size: usize, datatype: &Arc<Datatype>) -> Arc<RwLock<Varnode>> {
    let mut varnode = Varnode::new_constant(value, size);
    // Fixture input adapter: construct the exact Ghidra Varnode constructor
    // state independently of Rugra's separately tracked VARNODE-INIT wave.
    varnode.consumed = u64::MAX;
    varnode.nzm = value;
    varnode.v_type = Some(datatype.clone());
    Arc::new(RwLock::new(varnode))
}

fn run_copy(label: &str, size: usize, value: u64) {
    let datatype = Arc::new(Datatype::Base(TypeBase::new(
        "fixture_unknown".to_string(),
        size,
        TypeMetatype::Unknown,
    )));
    let input = constant(value, size, &datatype);
    let inputs = vec![input];
    let (output, op) = attach(size, 0x2000, OpCode::CPUI_COPY, 0x1000, &datatype, &inputs);
    dump_state(label, "before", &output, &op, &inputs);
    let result = gather_offset(&output);
    println!("{label}|result=0x{result:x}");
    dump_state(label, "after", &output, &op, &inputs);
}

fn run_add(label: &str, size: usize, left: u64, right: u64) {
    let datatype = Arc::new(Datatype::Base(TypeBase::new(
        "fixture_unknown".to_string(),
        size,
        TypeMetatype::Unknown,
    )));
    let input0 = constant(left, size, &datatype);
    let input1 = constant(right, size, &datatype);
    let inputs = vec![input0, input1];
    let (output, op) = attach(size, 0x2100, OpCode::CPUI_INT_ADD, 0x1100, &datatype, &inputs);
    dump_state(label, "before", &output, &op, &inputs);
    let result = gather_offset(&output);
    println!("{label}|result=0x{result:x}");
    dump_state(label, "after", &output, &op, &inputs);
}

fn main() {
    run_copy("copy8_all_bits", 8, 0xfedcba9876543210);
    run_add("add8_wrap", 8, 0xfffffffffffffff0, 0x35);
    run_add("add7_mask", 7, 0x00fffffffffffff0, 0x35);
}
